YOLOv5 ISICs Problem

RTongs opened the pull request:
This is a solution of identifying melanoma in the ISICs 2018 dataset using YOLOv5. I started out with Mask R-CNN but found it much too time consuming to try to convert it to TensorFlow 2. Maybe YOLOv3 or v1 would have been more appropriate, but I found this out too late. Most of my time was spent trying to do data augmentation, generating the masks, data splitting, and understanding how to use YOLOv5.

The technicalities of the data augmentation included rotations and flips. With the data splitting just being a standard 80/20 on training and test, and then validation taking 20% of the 80% of the training set.

I realise the importance of evaluating with a test set, but I was only able to test my only working model on just the validation set as I lost my model checkpoints and ran out of time. I did get the test inferencing working though, you are welcome to try it.

I was not able to fully finish my assignment due to unforeseen medical issues so apologies for the incomplete assignment. Even finishing the report was physically painful and damaging. Note that my username is ryemigie for one of my GitHub accounts which is why it came up. Apologies.

RTongs committed:
Bring in empty ReadME and initial driver code and model class.

RTongs committed:
Fix driver main calling and self in the model

RTongs committed:
Update README.md

Added an example README file from the web to get me started.

RTongsSizztech committed:
Bring in Mask R-CNN model

Use public GitHub project that converts original Mask R-CNN code written in TensorFlow 1 to TensorFlow 2.0. Bring in a sample image from the dataset, and add a .gitignore file to ignore model files. Bring in dataset in my model file. Bring support for program and default arguments in driver file.

RTongs committed:
Bug fixing TensorFlow 2 Mask R-CNN Code

Update all references to correctly use the keras stored in TensorFlow in newer versions. Fix bugs where model creation could not be performed due to the TrainingFlag not being specified on call() methods and tf.shape() was not being used.

RTongs committed:
Exclude Not Needed Layers

Change structure of model loaded in from hd5 file to exclude certain layers because we only have 2 classes instead of 81. This has allowed the model to finally run and give a (poor) prediction.

RTongs committed:
Refine Model Parameters, Begin Dataset Loading

Reduce some of the bounding box and other parameters by a factor of 10 to reflect our reduced number of classes versus the COCO dataset. Begin basic loading of our data into a training and validation set, with a test set to be done later (with care taken not to contaminate). Also doing a simple image flip augmentation. Next to do is create custom class with a load_image() and load_mask() method as required for Mask R-CNN.

RTongs committed:
Create Custom Dataset Class, Fix Further Bugs

Come up with a way of randomly selecting both the mask and the image and keeping them paired. Fix more bugs that are caused by newer versions of TensorFlow. Found most of the fixes by scouring GitHub for issues and bug reports on the original Mask R-CNN GitHub page.

RTongs committed:
Bug fix for reduce_mean()

Update call to tf.reduce_mean() to reflect change in TensorFlow API since 2.0.

RTongs committed:
Change from Mask R-CNN to YOLOv5

Remove all Mask R-CNN code and begin creation of simple YOLO class using PyTorch with the loaded medium sized model trained on COCO. Rename data loader class to setup class, that is now its own program with main. This program should be run to move the training data into subdirectories and randomly shuffle it. Additionally, we must create a YAML file and text files for each image to be able to pass our bounding boxes into the model, as YOLO does not use masks. Update .gitignore to not include Mac specific file and all of the training data in the subdirectories. Run the first inference, which bounded quite well but of course for the wrong class (cat).

RTongs committed:
Add Test Set Compatability

Enhance the isics_data_setup.py file to also generate a test set that has a size that can be specified as a runtime parameter. Add a runtime parameter to the model class to specify whether we do training or prediction.

RTongs committed:
Tweak Code in Preparation For Training Loop

Add print statements for the driver describing what mode it will run the YOLO model in. Update YOLO class to reflect differences from Mask R-CNN.

RTongs committed:
Implement Simple Training Loop

Make a custom training loop for YOLOv5. I don't think it can work, and I may have to use their train.py due to the bounding box architecture.

RTongs committed:
Update Data Loader, Begin Training

Update the data loader to put the mask info in a label folder instead of images, as required by YOLOv5. clone the YOLOv5 code into a yolov5 directory inside. Move our yaml file into the yolov5 directory.

RTongs committed:
Compress JPEGs to 1024

Resize all of the images with the maximum axis being 1024. This has brought the dataset size from 3.3GB to 170MB. Also ensure we are closing our files correctly by using the with operator, as file IO objects sit in unmanaged memory and will not be collected by the RC garbage collector.

RTongs committed:
Attach My Driver to YOLOv5 train.py

Replace my YOLO class for the YOLOv5 train.py file, as it is required because of all of the bells and whistles attached to YOLOv5. See my YOLO class for an example simple training loop in PyTorch. Add the necessary runtime parameter to my driver.py to accomodate YOLOv5.

RTongs committed:
Change Image Dimensions To 640, Augment Data

Change the resize default dimensions to 640 on the longest axis, which is the default for YOLOv5. Rewrite some of the isics_data_setup file to not copy the mask data into the new directories, as we don't need it, and also to generate augmented data. For each file we do a 90, 180, and 270 degrees rotation, as well as a horizontal and vertical flip from the original image.

RTongs committed:
Update README.md With Intended Layout

Just a first cut of the README file.

RTongs committed:
Reduce Training Data Generated

Too much training data was generated so change the isics_data_setup to randomly choose 2 of the 6 possible augmentations.

RTongs committed:
Remove misleading comment

Fix comment claiming the code will augment validation and test data, which it does not do anymore.

RTongs committed:
Update README

Bring in initial comments

RTongs committed:
Update README with first cut of info

Complete first cut of data processing and training methodology paragraphs.

RTongs committed:
Update README with Usage

Begin first simple cut of usage for both scripts.

RTongs committed:
Update for Test Set Evaluation

Use YOLOv5 detect.py to do basic inference.

RTongs committed:
Merge branch 'topic-recognition' of https://github.com/RTongs/PatternFlow into topic-recognition

RTongs committed:
Finalise README

Bring in basic results and YOLOv5 explanation.

RTongs committed:
Bring in validation images

Give some example inference images.

RTongs committed:
Important README update

Add a note explaining that user may need to just clone the YOLOv5 repository inside the s4427545_isics folder to use this repository. See the comment in README for more.

shakes76 commented:
## TF/Torch Usage
Adequate use and implementation
Late submission, but medical reasons
## Good Practice (Design/Commenting)
Good spacing and comments
No Header blocks -1
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present, but not applied to test data -1
No Data leakage 
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Please remove WandB files, these will clog up the repository. Your pull request will not be merged until they are removed. Feedback marks will be awarded when these are removed. -2
Request Description could be more informative (technical aspects i.e.) -1

RTongs commented:
Just updating the PR now.

RTongs committed:
Clean up repository

Remove wandb files

RTongs commented:
The merge conflicts is some weird bug I've been getting... it keeps mentioning recognition/ISICs_Unet/README.md which I have not touched at all. I'll try and fix now.

RTongs committed:
Fix one merge conflict

Revert changes I made at the very start of my project. Unsure why I did this...

RTongs committed:
Remove DS_Store file (macOS file)

Update .gitignore and remove unneeded files

RTongs committed:
Merge branch 'topic-recognition' into topic-recognition

RTongs commented:
Just made a minor update to my initial comment, and now I'm ready to merge. I deleted the wandb files too.

shakes76 commented:
Feedback marks awarded.

shakes76 commented:
Still can't merge though because of the .cache files present that take up some space.

RTongs committed:
Delete Cache Files

Remove the cache files and the JPG

RTongs committed:
Merge remote-tracking branch 'origin/topic-recognition' into topic-recognition

RTongs commented:
Deleted cache files, ready to merge.
